src/encode_url.rs

```diff
@@ -0,0 +1,114 @@
+  let mut i = 0;
+  while i < bytes.len() {
+    set = set.remove(bytes[i]);
+    i += 1;
+  }
+  set
+}
+
+// `percent_encoding::percent_encode` need a `&'static AsciiSet`.
+const ASCII_SET: &AsciiSet = &get_ascii_set();
+
+fn encoding_override(c: &str) -> Cow<[u8]> {
+  // decode, then encode
+  let dec = percent_decode_str(c).decode_utf8_lossy();
+  Cow::<str>::from(percent_encode(dec.as_bytes(), ASCII_SET))
+    .to_string()
+    .into_bytes()
+    .into()
+}
+
+fn encode_url_impl(input: &str) -> String {
+  let enc = Some(&encoding_override as _);
+  // Provide an `encoding_override` function to manually transform query part. Otherwise, url crate tends to leave
+  // percent signs as-is.
+  let mut parse_result = Url::options().encoding_override(enc).parse(input);
+  let mut is_relative = false;
+  if let Err(ParseError::RelativeUrlWithoutBase) = &parse_result {
+    is_relative = true;
+    // For relative URLs, we provide a short base URL which will be striped later.
+    let base = Url::parse("http://.").expect("http://. must be a valid URL");
+    parse_result = Url::options()
+      .base_url(Some(&base))
+      .encoding_override(enc)
+      .parse(input);
+  }
+  let url = match parse_result {
+    Ok(u) => u,
+    Err(_) => return input.to_owned(),
+  };
+  if is_relative {
+    let mut str = String::from(url);
+    if input.as_bytes()[0] == b'/' {
+      // Strip "http://."
+      str.drain(..8);
+    } else {
+      // An unnecessary leading slash was added to the URL during parsing.
+      str.drain(..9);
+    }
+    return str;
+  }
+  let domain = match url.domain() {
+    Some(d) => d,
+    None => return url.into(), // `javascript:` etc.
+  };
+  // The domain provided by url crate is punycode-encoded. We need to convert the domain back to unicode form, and copy
+  // over remaining URL parts. However, there seems no easy way to find what's before and after domain within a URL.
+  // Assuming the domain str slice is a part of the URL (inferred from elided lifetime on `url::Url::domain()`'s
+  // signature), we can obtain the relative byte position where domain starts within a URL by calculating the pointer
+  // difference between them, which will later be used to extract whatever surrounding the domain.
+
+  let url_str = url.as_str();
+  let url_ptr = url_str.as_ptr();
+  let domain_ptr = domain.as_ptr();
+  let domain_len = domain.len();
+  assert!(domain_len > 0);
+  assert!(domain_len <= url_str.len());
+  assert!(domain_ptr >= url_ptr);
+  assert!(
+    // Safety:
+    // 1) Since the domain part is guaranteed to be non-empty, `domain_ptr` should point to a valid allocated memory
+    //    address, and the resulting pointer will be exactly one byte past the end of the same string.
+    // 2) A `RawVec` guarantees never allocating more than isize::MAX bytes.
+    // 3) A `RawVec` guarantees length never overflows.
+    unsafe { domain_ptr.add(domain_len) } <=
```

Comment:
You can probably avoid this unsafe by using `as_ptr_range`.

Reply:
TIL🥰